editor/plugins/canvas_item_editor_plugin.cpp

```diff
@@ -2056,27 +2056,31 @@ bool CanvasItemEditor::_gui_input_move(const Ref<InputEvent> &p_event) {
 
 	if (drag_type == DRAG_MOVE || drag_type == DRAG_MOVE_X || drag_type == DRAG_MOVE_Y) {
 		// Move the nodes
-		if (m.is_valid()) {
+		if (m.is_valid() && !drag_selection.is_empty()) {
 			_restore_canvas_item_state(drag_selection, true);
 
 			drag_to = transform.affine_inverse().xform(m->get_position());
 			Point2 previous_pos;
-			if (!drag_selection.is_empty()) {
-				if (drag_selection.size() == 1) {
-					Transform2D xform = drag_selection[0]->get_global_transform_with_canvas() * drag_selection[0]->get_transform().affine_inverse();
-					previous_pos = xform.xform(drag_selection[0]->_edit_get_position());
-				} else {
-					previous_pos = _get_encompassing_rect_from_list(drag_selection).position;
-				}
+			if (drag_selection.size() == 1) {
+				Transform2D xform = drag_selection[0]->get_global_transform_with_canvas() * drag_selection[0]->get_transform().affine_inverse();
+				previous_pos = xform.xform(drag_selection[0]->_edit_get_position());
+			} else {
+				previous_pos = _get_encompassing_rect_from_list(drag_selection).position;
 			}
 
-			Point2 new_pos = snap_point(previous_pos + (drag_to - drag_from), SNAP_GRID | SNAP_GUIDES | SNAP_PIXEL | SNAP_NODE_PARENT | SNAP_NODE_ANCHORS | SNAP_OTHER_NODES, 0, nullptr, drag_selection);
-
-			if (drag_type == DRAG_MOVE_X) {
-				new_pos.y = previous_pos.y;
-			} else if (drag_type == DRAG_MOVE_Y) {
-				new_pos.x = previous_pos.x;
+			Point2 new_pos;
+			if (drag_selection.size() == 1 && (drag_type == DRAG_MOVE_X || drag_type == DRAG_MOVE_Y)) {
+				const CanvasItem *selected = drag_selection.front()->get();
+				if (drag_type == DRAG_MOVE_X) {
+					new_pos = Vector2(selected->get_local_mouse_position().x, 0);
+				} else {
+					new_pos = Vector2(0, selected->get_local_mouse_position().y);
+				}
+				new_pos = selected->get_transform().xform(new_pos);
+			} else {
+				new_pos = previous_pos + (drag_to - drag_from);
 			}
```

Comment:
Using `get_local_mouse_position()` seems hackish, no? :thinking:

AFAICT looking at that code `new_pos` is supposed to be in the coordinate space of the `selected`'s parent. So (in case it's preferred to not use `get_local_mouse_position` like that) I think all that needs to be done is something like this (hopefully it will just work, haven't tested or anything :smile:):

```suggestion
			Vector2 drag_delta = drag_to - drag_from;
			if (drag_selection.size() == 1 && (drag_type == DRAG_MOVE_X || drag_type == DRAG_MOVE_Y)) {
				// Drag by local axis, so transform the drag to local space, then zero an axis, then transform back to parent space.
				const CanvasItem *selected = drag_selection.front()->get();
				drag_delta = selected->get_transform().affine_inverse().basis_xform(drag_delta);
				if (drag_type == DRAG_MOVE_X) {
					drag_delta.y = 0;
				} else {
					drag_delta.x = 0;
				}
				drag_delta = selected->get_transform().basis_xform(drag_delta);
			}
			new_pos = previous_pos + drag_delta;
```

Reply:
It works, thx.
Using `get_local_mouse_position()` isn't really a hack, but your code would make it easier to remove the "reset" behavior I mentioned.